gix-object/src/tree/mod.rs

```diff
@@ -249,14 +249,57 @@ impl Ord for EntryRef<'_> {
     }
 }
 
+/// TODO
+/// Keep in mind that the path separator always is `/`, independent of the platform.
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct RepositoryPathPuf {
+    /// TODO
+    inner: BString,
+}
+
+impl std::fmt::Display for RepositoryPathPuf {
+    #[inline]
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.inner.fmt(f)
+    }
+}
+
+impl From<&str> for RepositoryPathPuf {
+    fn from(value: &str) -> Self {
+        Self { inner: value.into() }
+    }
+}
+
+impl From<&BStr> for RepositoryPathPuf {
+    fn from(value: &BStr) -> Self {
+        Self { inner: value.into() }
+    }
+}
+
+impl From<BString> for RepositoryPathPuf {
+    fn from(value: BString) -> Self {
+        Self { inner: value }
+    }
+}
+
+impl std::ops::Deref for RepositoryPathPuf {
```

Comment:
The idea would be to avoid having easy conversions to `BString` and treat it as distinct type with its own PathBuf-like API.

Reply:
Absolutely! This was just a shortcut to get this first addition to compile with the least amount of changes possible (I should have been more clear about the context and the status of this initial commit).

A more general question: is the idea to have both the new types backed by `inner: BString` (and `inner: &BStr`) or rather by `inner: Vec<u8>` (and `inner: &[u8]`)? If the latter, we would need to implement all the necessary functions that are currently provided by `BString` (and `&BStr`) ourselves, or am I missing something?

Reply:
The path of least resistance (and the preferred one) should be to let it be backed by `BString|&BStr`. Maybe this will also be a great preparation for the time when `std::…::ByteString` can be used, when only the backing needs to be adjusted as most code is using `RepositoryPathComponent|RepositoryPathBuf` and friends.